### PR TITLE
Make our testing infrastructure resource-aware

### PR DESCRIPTION
**Pull request by bartgol**
There are three components to this:

1) EkatCreateUnitTest now requests resources for each test. In particular, it requests N "devices", where N is the number of ranks. The keyword "devices" is completely arbitrary for CTest. All CTest does is to then look for a resource spec json file (passed to ctest via `--resource-spec-file my_file.json`, and, if CTEST_PARALLEL_LEVEL>1, schedule tests based on the resources requested, and the resources available.

2) test-all-scream creates a json file on the fly for each build type, where the number of available devices is taken from `self._testing_res_count[test]`. This quantities are computed as the CTEST_PARALLEL_LEVEL divided by the number of builds (some build get extra testing resources if there's a remainder). On CPU, this is the number of cores used by each build. On GPU, this is the number of GPUs that each build could use. The id of the devices are offset, so that two builds have different ids. CTest does not care what we use as id, and for CPU builds it doesn't matter at all. But for GPU build, this id is important for us.
Notice that, up to now, testing_res_count on GPU was kind of pointless, since all tests were funneled to the same GPU.

3) Finally, in ekat we added a testing utility, which retrieves the device id for the current rank. For non-GPU builds, this number is not even computed, and nothing changes. test-all is already able to place different builds on different cores, and ctest can already place different tests (within the same build) on different cores (if possible). On GPU builds, however, we use this device_id to tell Kokkos which gpu we should use. The device id is retrieved from some env variable that ctest sets before invoking for each test, which are taken from the test properties set during the call to `EkatCreateUnitTest`. Notice that, if you don't pass a valid resource spec file, ctest will simply "ignore" the resource groups property set on the test.

Note: the resource specification feature is relatively new in CMake/CTest. I believe it was introduced in 3.16. However, setting the resource property is completely innocuous for previous versions (it will simply be unused). This means that this feature does _not_ require you to update your CMake/CTest versions.

Note: I haven't yet checked this on GPU (I am too tired now, so I'll let the AT test for me). But I manually hacked a `CTestTestFile.cmake` file, requesting for "too many" slots on a particular resource, and verified that CTest does not run the test, reporting the error message:

```
  Test requested resources of type 'devices' in the following amounts:
    2 slots
  but only the following units were available:
    '6': 1 slot
    '7': 1 slot
```

Note: if you're really into this, you might wonder why I only allow 1 slot. We can chat about that if you're interested. I thought using 1 slot per resource group (1 group=1 core on CPU, 1 group=1 gpu on GPU) made things much easier.

### CODE REVIEW

File: components/scream/ekat/src/ekat/util/ekat_string_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#ifndef EKAT_STRING_UTILS_HPP
+#define EKAT_STRING_UTILS_HPP
+
+#include <string>
+#include <vector>
+
```
Comment (jgfouca) [inline components/scream/ekat/src/ekat/util/ekat_string_utils.hpp:6]: Could you provide a paragraph comment here describing the higher-level idea behind this header. What is provides and why it is needed?

File: components/scream/ekat/src/ekat/util/scream_test_utils.hpp [hunk 2]
```diff
@@ -73,6 +72,8 @@ void catch2_req_pk_sensitive(const Scalar lhs, const Scalar rhs)
   }
 }
 
+int get_test_device (const int mpi_rank);
```
Comment (jgfouca) [inline components/scream/ekat/src/ekat/util/scream_test_utils.hpp:75]: This seems like it encapsulates a concept that is now very critical to our testing process. Could you provide some comments about this here?

File: components/scream/scripts/test_all_scream.py [hunk 3]
```diff
@@ -331,6 +332,40 @@ def get_taskset_id(self, test):
 
         return start, end
 
+    ###############################################################################
+    def create_ctest_resource_file(self, test, build_dir):
+    ###############################################################################
+        # Create a json file in the test build dir, which ctest will then use
```
Comment (jgfouca) [inline components/scream/scripts/test_all_scream.py:338]: Great comment, thanks

File: components/scream/ekat/src/ekat/util/scream_test_utils.hpp (hunk 2 above)
Comment (bartgol) [reply components/scream/ekat/src/ekat/util/scream_test_utils.hpp:100]: done

File: components/scream/ekat/src/ekat/util/ekat_string_utils.hpp (hunk 1 above)
Comment (bartgol) [reply components/scream/ekat/src/ekat/util/ekat_string_utils.hpp:6]: done

**Pull request merged**